README.md
# ProgressStepView
A library for creating progress with steps.

<p align="center" >
<img src="https://github.com/gustavoSAS/ProgressStepView/blob/master/Screenshots/increasing.gif" width="50%" height="50%" />
</p>

<p align="center" >
<img src="https://github.com/gustavoSAS/ProgressStepView/blob/master/Screenshots/decreasing.gif" width="50%" height="50%" />
</p>


## Installation

### CocoaPods

[CocoaPods](https://cocoapods.org) is a dependency manager for Cocoa projects. For usage and installation instructions, visit their website. To integrate ProgressStepView into your Xcode project using CocoaPods, specify it in your `Podfile`:

```ruby
pod 'ProgressStepView', '~> 1.0.0'
```


## Requirements

- iOS 10.0 or later
- Xcode 10.0 or later


## How To Use

* Swift

```swift
import ProgressStepView

private lazy var progress: ProgressStepView = {
    let progress = ProgressStepView(frame: viewTeste.bounds)
    progress.numberOfPoints = 4
    progress.spacing = 5.0
    progress.progressLineHeight = 8
    progress.circleRadius = 20
    progress.circleStrokeWidth = 2.5
    progress.translatesAutoresizingMaskIntoConstraints = false

    return progress
}()

view.addSubview(progress)

NSLayoutConstraint.activate([
    progress.topAnchor.constraint(equalTo: view.topAnchor, constant: 100),
    progress.leadingAnchor.constraint(equalTo: view.leadingAnchor),
    progress.trailingAnchor.constraint(equalTo: view.trailingAnchor)
])

progress.progress = 0.5
```

### Customization
Values of following properties have been set as defaults already. Change them if they are not suitable for you.

```swift
progress.numberOfPoints = 4
progress.spacing = 5.0
progress.progressLineHeight = 8
progress.progressCornerRadius = 4
progress.circleRadius = 20
progress.circleStrokeWidth = 2.5
progress.circleTintColor = #colorLiteral(red: 0, green: 0.4784313725, blue: 1, alpha: 1)
progress.circleColor = #colorLiteral(red: 0.6000000238, green: 0.6000000238, blue: 0.6000000238, alpha: 1)
progress.descriptionSpacing = 10.0
progress.descriptionColor = .black
progress.descriptionFont = UIFont.systemFont(ofSize: 15, weight: .bold)
progress.position = .top
```

### Add Description
Implement delegate, choose position

```swift
progress.position = .bottom
```
<p align="center" >
<img src="https://github.com/gustavoSAS/ProgressStepView/blob/master/Screenshots/description_below.png" width="50%" height="50%" />
</p>

```swift
progress.position = .top
```

<p align="center" >
<img src="https://github.com/gustavoSAS/ProgressStepView/blob/master/Screenshots/description above.png" width="50%" height="50%" />
</p>

```swift
progress.delegate = self
progress.position = .bottom

...

extension ViewController: ProgressStepViewDelegate {
    func progressStepView(descriptionForRow row: Int) -> String? {
        return "Number \(row)"
    }
}
```
## License

ProgressStepView is released under the MIT license. [See LICENSE](https://github.com/gustavoSAS/ProgressStepView/blob/master/LICENSE) for details.


CHANGELOG.md
# ProgressStepView
All notable changes to this project will be documented in this file.

## [1.0.0](https://github.com/gustavoSAS/ProgressStepView/releases/tag/1.0.0)
Released on 2019-08-10.

#### Added
- Initial release of ProgressStepView.
- Added by [Gustavo Storck](https://github.com/gustavoSAS).


ProgressStepView/Source/StepView.swift
//
//  StepView.swift
//  ProgressStep
//
//  Created by Gustavo Storck on 03/06/19.
//  Copyright © 2019 Gustavo Storck. All rights reserved.
//

import UIKit

class StepView: UIView {
    // MARK: Private attributes
    private lazy var checkImage: UIImageView = {
        let bundle = Bundle(for: StepView.self)
        guard let url = bundle.url(forResource: "ProgressStepView", withExtension: "bundle") else { return UIImageView() }
        
        let image = UIImage(named: "ic_check", in: Bundle(url: url), compatibleWith: nil)
        let imageView = UIImageView(image: image)
        imageView.alpha = 0.0
        imageView.tintColor = .white
        imageView.translatesAutoresizingMaskIntoConstraints = false
        
        return imageView
    }()
    
    private lazy var stepView: UIView = {
        let view = UIView()
        view.clipsToBounds = true
        view.layer.cornerRadius = circleRadius / 2
        view.layer.borderWidth = circleStrokeWidth
        view.backgroundColor = .clear
        view.layer.borderColor = circleColor.cgColor
        view.translatesAutoresizingMaskIntoConstraints = false
        
        return view
    }()
    
    private lazy var descStepView: UIView = {
        let view = UIView()
        view.clipsToBounds = true
        view.layer.cornerRadius = descCircleRadius / 2
        view.backgroundColor = circleTintColor
        view.translatesAutoresizingMaskIntoConstraints = false
        view.isHidden = true
        
        return view
    }()
    
    private var descCircleRadius: CGFloat {
        return circleRadius / 2
    }
    
    // MARK: Public attributes
    
    /// The size of circle
    var circleRadius: CGFloat = 20
    
    /// The width of the circle border
    var circleStrokeWidth: CGFloat = 2.5
    
    /// The color shown for the portion of the circle that is not filled.
    var circleColor: UIColor = .gray
    
    /// The color shown for the portion of the circle that is filled.
    var circleTintColor: UIColor = .blue
    
    /// The animation time, between the state transitions.
    var animationTime: TimeInterval = 0.3
    
    // MARK: Life cycle
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override public func didMoveToSuperview() {
        super.didMoveToSuperview()
        
        addComponents()
        layoutComponents()
    }
    
    // MARK: Private functions
    
    private func addComponents() {
        addSubview(stepView)
        addSubview(descStepView)
        addSubview(checkImage)
    }
    
    private func layoutComponents() {
        NSLayoutConstraint.activate([
            stepView.topAnchor.constraint(equalTo: topAnchor),
            stepView.bottomAnchor.constraint(equalTo: bottomAnchor),
            stepView.leadingAnchor.constraint(equalTo: leadingAnchor),
            stepView.trailingAnchor.constraint(equalTo: trailingAnchor),
            stepView.widthAnchor.constraint(equalToConstant: circleRadius),
            stepView.heightAnchor.constraint(equalToConstant: circleRadius)
        ])
        
        NSLayoutConstraint.activate([
            descStepView.centerXAnchor.constraint(equalTo: stepView.centerXAnchor),
            descStepView.centerYAnchor.constraint(equalTo: stepView.centerYAnchor),
            descStepView.widthAnchor.constraint(equalToConstant: descCircleRadius),
            descStepView.heightAnchor.constraint(equalToConstant: descCircleRadius)
        ])
        
        NSLayoutConstraint.activate([
            checkImage.topAnchor.constraint(equalTo: topAnchor),
            checkImage.bottomAnchor.constraint(equalTo: bottomAnchor),
            checkImage.leadingAnchor.constraint(equalTo: leadingAnchor),
            checkImage.trailingAnchor.constraint(equalTo: trailingAnchor)
        ])
    }
    
    private func animatePending() {
        UIView.animate(withDuration: animationTime, delay: 0.0, options: [.curveEaseOut], animations: { [weak self] in
            self?.descStepView.alpha = 0.0
            self?.checkImage.alpha = 0.0
            self?.stepView.backgroundColor = .clear
            self?.stepView.layer.borderWidth = self?.circleStrokeWidth ?? 0.0
            self?.stepView.transform = CGAffineTransform(scaleX: 1.0, y: 1.0)
        }, completion: { [weak self] _ in
            self?.descStepView.isHidden = true
        })
    }
    
    private func animateWait() {
        descStepView.alpha = 0.0
        descStepView.isHidden = false
        UIView.animate(withDuration: animationTime, delay: 0, options: [.curveEaseOut], animations: { [weak self] in
            self?.descStepView.alpha = 1.0
            self?.checkImage.alpha = 0.0
            self?.stepView.backgroundColor = .clear
            self?.stepView.layer.borderWidth = self?.circleStrokeWidth ?? 0.0
            self?.stepView.transform = CGAffineTransform(scaleX: 1.0, y: 1.0)
        })
    }
    
    private func animateComplete() {
        UIView.animate(withDuration: animationTime, delay: 0.0, options: [.curveEaseOut], animations: { [weak self] in
            self?.checkImage.alpha = 1.0
            self?.stepView.backgroundColor = self?.circleTintColor
            self?.stepView.layer.borderWidth = 0.0
            self?.stepView.transform = CGAffineTransform(scaleX: 1.05, y: 1.05)
        }, completion: { [weak self] _ in
            self?.descStepView.isHidden = false
            self?.descStepView.alpha = 1.0
        })
    }
    
    // MARK: Public functions
    
    /// Performs a state transition animation
    ///
    /// - Parameter status: Indicates the next StepView status
    func animateTo(_ status: Status) {
        switch status {
        case .pending:
            animatePending()
        case .wait:
            animateWait()
        case .complete:
            animateComplete()
        }
    }
}

// MARK: Extensions

extension StepView {
    /// Status that StepView can take
    ///
    /// - pending: It does not show any marking
    /// - wait: Displays the current status mark
    /// - complete: Displays full status marking
    enum Status {
        case pending
        case wait
        case complete
    }
}


Example/ProgressStepViewDemo/AppDelegate.swift
//
//  AppDelegate.swift
//  ProgressStepViewDemo
//
//  Created by Gustavo Storck on 09/08/19.
//  Copyright © 2019 Gustavo Storck. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



ProgressStepView/Source/ProgressStepView.swift
//
//  ProgressStepView.swift
//  ProgressStep
//
//  Created by Gustavo Storck on 03/06/19.
//  Copyright © 2019 Gustavo Storck. All rights reserved.
//

import UIKit
// MARK: Delegate

public protocol ProgressStepViewDelegate: class {
    /// Waits for the description text of the current step
    ///
    /// - Parameter row: Represents the current step view position
    /// - Returns: Description text that will be assigned to step view
    func progressStepView(descriptionForRow row: Int) -> String?
}

public class ProgressStepView: UIView {
    // MARK: Private attributes
    private var progressViewArray: [UIProgressView] = []
    private var stepViewArray: [StepView] = []
    private var descriptionLabelArray: [(UILabel, Int)] = []
    private var currentProgress: Float = 0
    private var animateFinish: Bool = true
    
    private lazy var stackView: UIStackView = {
        let stackView = UIStackView()
        stackView.alignment = .center
        stackView.distribution = .fill
        stackView.axis = .horizontal
        stackView.spacing = spacing
        stackView.translatesAutoresizingMaskIntoConstraints = false
        
        return stackView
    }()
    
    private var progressView: UIProgressView {
        let progressView = UIProgressView()
        progressView.trackTintColor = circleColor
        progressView.progressTintColor = circleTintColor
        progressView.progress = 0.0
        progressView.translatesAutoresizingMaskIntoConstraints = false
        
        return progressView
    }
    
    private var stepView: StepView {
        let stepView = StepView()
        stepView.circleRadius = circleRadius
        stepView.circleStrokeWidth = circleStrokeWidth
        stepView.circleColor = circleColor
        stepView.circleTintColor = circleTintColor
        stepView.translatesAutoresizingMaskIntoConstraints = false
        
        return stepView
    }
    
    private var descriptionLabel: UILabel {
        let label = UILabel()
        label.textAlignment = .center
        label.font = descriptionFont
        label.textColor = descriptionColor
        label.numberOfLines = 0
        label.translatesAutoresizingMaskIntoConstraints = false
        
        return label
    }
    
    // MARK: Public attributes
    
    /// Number of progress steps
    public var numberOfPoints: Int = 2 {
        didSet {
            numberOfPoints = numberOfPoints < 2 ? 2 : numberOfPoints
        }
    }
    
    /// Represents the height of the progreeView
    public var progressLineHeight: CGFloat = 5
    
    /// The radius to use when drawing rounded corners ProgressLine
    public var progressCornerRadius: CGFloat = 0
    
    /// The size of the space between a step and a progress view
    public var spacing: CGFloat = 5.0
    
    /// The size of circle
    public var circleRadius: CGFloat = 20
    
    /// The width of the circle border
    public var circleStrokeWidth: CGFloat = 2.5
    
    /// The color shown for the portion of the circle that is not filled.
    public var circleColor: UIColor = .gray
    
    /// The color shown for the portion of the circle that is filled.
    public var circleTintColor: UIColor = .blue
    
    /// Description text position.
    ///
    /// - top: Text over the step
    /// - bottom: Text below step
    public var position: Position = .top
    
    /// The font used to display the text.
    public var descriptionFont: UIFont = UIFont.systemFont(ofSize: 14, weight: .regular)
    
    /// The color of the text.
    public var descriptionColor: UIColor = .black
    
    /// Space between description text and stepView.
    public var descriptionSpacing: CGFloat = 10.0
    
    /// The current progress shown by the receiver.
    ///
    /// The current progress is represented by a floating-point value
    /// between 0.0 and 1.0, inclusive, where 1.0 indicates the completion
    /// of the task. The default value is 0.0. Values less than 0.0 and greater than 1.0
    /// are pinned to those limits.
    public var progress: Float = 0.0 {
        didSet {
            progress = progress > 1 ? 1 : progress
            progress = progress < 0 ? 0 : progress
            
            updateProgress(to: progress)
        }
    }
    
    public weak var delegate: ProgressStepViewDelegate?
    
    // MARK: Life cycle
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
    }
    
    public override func didMoveToSuperview() {
        super.didMoveToSuperview()
        
        addComponents()
        addDescription()
        layoutComponents()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: Private functions
    
    private func addComponents() {
        for _ in 0..<(numberOfPoints - 1) {
            let step = stepView
            let progress = progressView
            
            stackView.addArrangedSubview(step)
            stackView.addArrangedSubview(progress)
            
            stepViewArray.append(step)
            progressViewArray.append(progress)
        }
        
        let step = stepView
        stepViewArray.append(step)
        stackView.addArrangedSubview(step)
        addSubview(stackView)
    }
    
    private func addDescription() {
        for index in 0..<stepViewArray.count {
            guard let text = delegate?.progressStepView(descriptionForRow: index) else { continue }
            let stepView = stepViewArray[index]
            let label = descriptionLabel
            
            label.text = text
            addSubview(label)
            layoutDescription(label, stepView: stepView)
            descriptionLabelArray.append((label, index))
        }
    }
    
    private func layoutDescription(_ label: UILabel, stepView: StepView) {
        let constraintLabel = constraint(label: label, stepView: stepView)
        let width = valueWidthDescription()
        
        NSLayoutConstraint.activate([
            label.centerXAnchor.constraint(equalTo: stepView.centerXAnchor),
            constraintLabel,
            label.widthAnchor.constraint(lessThanOrEqualToConstant: width)
        ])
    }
    
    private func constraint(label: UILabel, stepView: StepView) -> NSLayoutConstraint {
        switch position {
        case .top:
            return label.bottomAnchor.constraint(equalTo: stepView.topAnchor, constant: -descriptionSpacing)
        case .bottom:
            return label.topAnchor.constraint(equalTo: stepView.bottomAnchor, constant: descriptionSpacing)
        }
    }
    
    private func layoutComponents() {
        switch position {
        case .bottom:
            layoutComponentsBottom()
        case .top:
            layoutComponentsTop()
        }
        
        if let label = startLabel(), label.intrinsicContentSize.width >= circleRadius {
            label.leadingAnchor.constraint(equalTo: leadingAnchor).isActive = true
        } else {
            stackView.leadingAnchor.constraint(equalTo: leadingAnchor).isActive = true
        }
        
        if let label = endLabel(), label.intrinsicContentSize.width >= circleRadius {
            label.trailingAnchor.constraint(equalTo: trailingAnchor).isActive = true
        } else {
            stackView.trailingAnchor.constraint(equalTo: trailingAnchor).isActive = true
        }
        
        for progress in progressViewArray {
            configureProgressView(progressView: progress)
        }
    }
    
    private func configureProgressView(progressView: UIProgressView) {
        NSLayoutConstraint.activate([
            progressView.widthAnchor.constraint(equalTo: progressViewArray[0].widthAnchor),
            progressView.heightAnchor.constraint(equalToConstant: progressLineHeight)
        ])
        
        progressView.clipsToBounds = true
        progressView.layer.cornerRadius = progressCornerRadius
    }
    
    private func layoutComponentsTop() {
        if let label = maxHeightLabel() {
            label.topAnchor.constraint(equalTo: topAnchor).isActive = true
        } else {
            stackView.topAnchor.constraint(equalTo: topAnchor).isActive = true
        }
        
        stackView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
    }
    
    private func layoutComponentsBottom() {
        if let label = maxHeightLabel() {
            label.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
        } else {
            stackView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
        }
        
        stackView.topAnchor.constraint(equalTo: topAnchor).isActive = true
    }
    
    // MARK: Calculate functions
    
    private func valueWidthProgress() -> CGFloat {
        let numberSpacing = 2 * (CGFloat(numberOfPoints) - 1) * spacing
        let width = (frame.width - (circleRadius * CGFloat(numberOfPoints) + numberSpacing)) / CGFloat(numberOfPoints - 1)
        
        return width
    }
    
    private func valueWidthDescription() -> CGFloat {
        let widthProgress = 0.6 * valueWidthProgress()
        let spacingWidth = 2 * spacing
        let stepWidth = circleRadius
        
        return (widthProgress + spacingWidth + stepWidth)
    }
    
    private func startLabel() -> UILabel? {
        var startLabel: UILabel?
        for (label, index) in descriptionLabelArray {
            if index == 0 { startLabel = label }
        }
        return startLabel
    }
    
    private func endLabel() -> UILabel? {
        var endLabel: UILabel?
        for (label, index) in descriptionLabelArray {
            if index == (numberOfPoints - 1) { endLabel = label }
        }
        return endLabel
    }
    
    private func maxHeightLabel() -> UILabel? {
        var maxLabel: UILabel?
        for (label, _) in descriptionLabelArray {
            if label.intrinsicContentSize.width >= (maxLabel?.intrinsicContentSize.width ?? 0) {
                maxLabel = label
            }
        }
        
        return maxLabel
    }
    
    // MARK: Animate functions
    
    private func updateProgress(to progress: Float) {
        guard animateFinish else { return }
        let numStep = Float(numberOfPoints - 1)
        let value = progress * numStep
        
        animateFinish = false
        if value < currentProgress {
            animateDown(index: progressViewArray.endIndex - 1, value: numStep - value)
        } else {
            animateUp(index: 0, value: value)
        }
        
        currentProgress = value
    }
    
    private func animateDown(index: Int, value: Float) {
        let x = value - 1 < 0 ? -(value - 1) : 0
        
        if index == (progressViewArray.startIndex - 1) || x >= 1 {
            if x == 1 {
                stepViewArray[index + 1].animateTo(.wait)
            }
            animateFinish = true
            return
        }
        
        let progressView = progressViewArray[index]
        
        if x < 1 {
            stepViewArray[index + 1].animateTo(.pending)
        } else if x == 1 {
            stepViewArray[index + 1].animateTo(.wait)
        }
        
        let time: Float = progressView.progress - x
        progressView.setProgress(x, animated: false)
        UIView.animate(withDuration: (TimeInterval(time * 0.8)), animations: {
            progressView.layoutIfNeeded()
        }, completion: { [weak self] _ in
            self?.animateDown(index: index - 1, value: value - 1)
        })
    }
    
    private func animateUp(index: Int, value: Float) {
        let x = value - 1 < 0 ? value : 1
        if index == progressViewArray.endIndex || x < 0 {
            if x == 0 {
                stepViewArray[index].animateTo(.complete)
            }
            animateFinish = true
            return
        }
        
        let progressView = progressViewArray[index]
        
        if x > 0 {
            stepViewArray[index].animateTo(.complete)
        } else if x == 0 {
            stepViewArray[index].animateTo(.wait)
        }
        
        let time: Float = x - progressView.progress
        progressView.setProgress(x, animated: false)
        UIView.animate(withDuration: TimeInterval(time * 0.8), animations: {
            progressView.layoutIfNeeded()
        }, completion: { [weak self] _ in
            self?.animateUp(index: index + 1, value: value - 1)
        })
    }
}

// MARK: Extensions

extension ProgressStepView {
    /// Represents the position of the text in relation to the view.
    ///
    /// - top: Text over the step
    /// - bottom: Text below step
    public enum Position {
        case top
        case bottom
    }
}


Example/ProgressStepViewDemo/ViewController.swift
//
//  ViewController.swift
//  ProgressStepViewDemo
//
//  Created by Gustavo Storck on 09/08/19.
//  Copyright © 2019 Gustavo Storck. All rights reserved.
//

import UIKit
import ProgressStepView

class ViewController: UIViewController {
    @IBOutlet weak var slider: UISlider!
    
    private lazy var progress: ProgressStepView = {
        let progress = ProgressStepView(frame: view.bounds)
        progress.numberOfPoints = 4
        progress.spacing = 5.0
        progress.progressLineHeight = 8
        progress.progressCornerRadius = 4
        progress.circleRadius = 20
        progress.circleStrokeWidth = 2.5
        progress.circleTintColor = #colorLiteral(red: 0, green: 0.4784313725, blue: 1, alpha: 1)
        progress.circleColor = #colorLiteral(red: 0.6000000238, green: 0.6000000238, blue: 0.6000000238, alpha: 1)
        progress.descriptionSpacing = 10.0
        progress.descriptionColor = .black
        progress.descriptionFont = UIFont.systemFont(ofSize: 15, weight: .bold)
        progress.position = .bottom

        progress.translatesAutoresizingMaskIntoConstraints = false
        progress.delegate = self

        return progress
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.addSubview(progress)

        NSLayoutConstraint.activate([
            progress.topAnchor.constraint(equalTo: view.topAnchor, constant: 100),
            progress.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 10),
            progress.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -10)
        ])
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        progress.progress = 0.5
    }

    @IBAction func didTapButton(_ sender: Any) {
        progress.progress = slider.value
    }
}

extension ViewController: ProgressStepViewDelegate {
    func progressStepView(descriptionForRow row: Int) -> String? {
        return "Number \(row)"
    }
}
